### ISSUE TITLE
Java heap memory stats by OSHI?

### ISSUE THREAD

**Issue by pwankhed** [oshi/oshi#1014]
Hello Experts,

Is there any way OSHI can return the Java Heap usage details? 

I've created OSProcess p; and p is returning the following,

String Mem = p.getVirtualSize() + ""; but did not find heap details? Any hint or example would be very much appreciated.

**Comment by dbwiddis** [comment]
Short answer: no.  :)

Longer answer:  heap usage is a rapidly changing statistic and by the time you've measured it, it's changed.  In Java, particularly, it is allowed to grow toward a max value and the recovered (mostly) via garbage collection.   Java releases usually provide various methods of measuring the heap memory of java processes (e.g., jconsole, visualvm, jstat, and others) and most software profilers will provide this information as well -- or even better, a detailed rundown of exactly what objects are in the heap -- generally a more useful measure.

Bottom line: in theory we could try to measure and report an overall number, but I doubt it would have any practical value.

**Comment by andrm** [comment]
@pwankhed The information is also available in the following classes: 
java.lang.management.MemoryMxBean / GarbageCollectorMBean
java.lang.Runtime

**Comment by dbwiddis** [comment]
That's only for the running process. I think the intent was to monitor other processes.

**Comment by andrm** [comment]
Ah sorry, I read that wrong. I generally agree with your answer.

On the other hand, thinking about other processes: it could be useful though. I use oshi to gather system information and report it to a central point. If I need information about other java processes it would be nice to see the heap info. Elastic Search is notorious for using all the heap and even running out.

If oshi is the central component for gathering this information maybe it is the right place to read out the jvm memory info also. The API that is used by jstat/jconsole could be used.

**Comment by pwankhed** [comment]
Thanks @dbwiddis & @andrm for your help.

I agree with @andrm. We are writing an application which is expected to be a single point of source to collect the system info, including the JVM stats. This tool will be used on a production system wherein the other Java monitoring tools i.e. JVisualVM or JConsole cannot be installed. Our enterprise software is a Java based tool and our core engineering team will be certifying this new tool to collect the JVM information from customer's machine. I agree that heap usage are changed in micro seconds but our idea is to keep this tool running in the background and collect the details say after every 2-3 seconds etc. It would be nice if this feature is made available.

**Comment by dbwiddis** [comment]
If I understand @andrm correctly, the feature "is available" in core Java using the MXBeans.  We don't need to duplicate that functionality with OSHI unless we are adding value to it somehow.
